uniprotkb-rest/src/main/java/org/uniprot/api/uniprotkb/controller/UniProtKBDownloadController.java

```diff
@@ -0,0 +1,146 @@
+import org.springframework.web.bind.annotation.*;
+import org.uniprot.api.common.concurrency.Gatekeeper;
+import org.uniprot.api.common.repository.search.ProblemPair;
+import org.uniprot.api.rest.controller.BasicSearchController;
+import org.uniprot.api.rest.download.model.DownloadJob;
+import org.uniprot.api.rest.download.model.DownloadRequestToArrayConverter;
+import org.uniprot.api.rest.download.model.HashGenerator;
+import org.uniprot.api.rest.download.model.JobStatus;
+import org.uniprot.api.rest.download.queue.ProducerMessageService;
+import org.uniprot.api.rest.download.repository.DownloadJobRepository;
+import org.uniprot.api.rest.output.context.MessageConverterContextFactory;
+import org.uniprot.api.rest.output.job.DownloadJobDetailResponse;
+import org.uniprot.api.rest.output.job.JobStatusResponse;
+import org.uniprot.api.rest.output.job.JobSubmitResponse;
+import org.uniprot.api.rest.request.DownloadRequest;
+import org.uniprot.api.uniprotkb.controller.request.UniProtKBDownloadRequest;
+import org.uniprot.core.uniprotkb.UniProtKBEntry;
+
+import io.swagger.v3.oas.annotations.Operation;
+import io.swagger.v3.oas.annotations.media.Content;
+import io.swagger.v3.oas.annotations.media.Schema;
+import io.swagger.v3.oas.annotations.responses.ApiResponse;
+
+/**
+ * @author sahmad
+ * @created 21/11/2022
+ */
+@RestController
+@RequestMapping(value = DOWNLOAD_RESOURCE)
+public class UniProtKBDownloadController extends BasicSearchController<UniProtKBEntry> {
+    static final String DOWNLOAD_RESOURCE = UNIPROTKB_RESOURCE + "/download";
+    private final ProducerMessageService messageService;
+    private final DownloadJobRepository jobRepository;
+    private final HashGenerator<DownloadRequest> hashGenerator;
+    public static final String JOB_ID = "jobId";
+
+    private static final String SALT_STR = "UNIPROT_DOWNLOAD_SALT";
+
+    public UniProtKBDownloadController(
+            ProducerMessageService messageService,
+            DownloadJobRepository jobRepository,
+            ApplicationEventPublisher eventPublisher,
+            MessageConverterContextFactory<UniProtKBEntry> converterContextFactory,
+            ThreadPoolTaskExecutor downloadTaskExecutor,
+            Gatekeeper downloadGatekeeper) {
+        super(
+                eventPublisher,
+                converterContextFactory,
+                downloadTaskExecutor,
+                UNIPROTKB,
+                downloadGatekeeper);
+        this.messageService = messageService;
+        this.jobRepository = jobRepository;
+        this.hashGenerator = new HashGenerator<>(new DownloadRequestToArrayConverter(), SALT_STR);
+    }
+
+    @PostMapping(value = "/run", produces = APPLICATION_JSON_VALUE)
+    public ResponseEntity<JobSubmitResponse> submitJob(
+            @Valid @ModelAttribute UniProtKBDownloadRequest request,
+            HttpServletRequest httpRequest) {
+        MessageProperties messageHeader = new MessageProperties();
```

Comment:
IMO, I would prefer an intermediate class to do these kinds of processing. It has happened because you directly talk to the repository from the controller itself so that you dont have any other place to write the processing logic.